README.md
# isoweek

<p align="center">
<a href="https://pub.dev/packages/isoweek"><img src="https://img.shields.io/pub/v/isoweek?color=green" alt="pub.dev badge"></a>
<a href="https://github.com/devj3ns/isoweek/commits/main"><img alt="GitHub commit activity" src="https://img.shields.io/github/commit-activity/m/devj3ns/isoweek?color=blue&label=commits"></a>
<br>
<a href="https://www.buymeacoffee.com/devj3ns" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-blue.png" alt="Buy Me A Coffee" height="42px"></a>
</p>

A lightweight Dart package which provides the class Week (based on the local time). Instances represent specific weeks spanning Monday to Sunday. 
An ISO week-numbering year has 52 or 53 full weeks. Week 1 is defined to be the first week with 4 or more days in January.

It’s called isoweek because this is the week definition of [ISO 8601](https://en.wikipedia.org/wiki/ISO_week_date).
This package is inspired by the [isoweek Python package](https://pypi.org/project/isoweek/).

### Example

```dart
import 'package:isoweek/isoweek.dart';

void main() {
  Week currentWeek = Week.current();
  print('Current week: $currentWeek');
  //OUTPUT: 2021W13

  print('Days of current week: ${currentWeek.days}');
  //OUTPUT: [2021-03-29 00:00:00.000, 2021-03-30 00:00:00.000, ...]

  Week nextWeek = currentWeek.next;
  print('Days of next week: ${nextWeek.days}');
  //OUTPUT: Days of next week: [2021-04-05 00:00:00.000, 2021-04-06 00:00:00.000, ...]

  Week previousWeek = currentWeek.previous;
  print('Days of previous week: ${previousWeek.days}');
  //OUTPUT: Days of previous week: [2021-03-22 00:00:00.000, 2021-03-23 00:00:00.000, ...]

  Week futureWeek = currentWeek.addWeeks(5);
  print('5 weeks ahead: $futureWeek');
  //OUTPUT: 5 weeks ahead: 2021W18

  Week birthdayWeek = Week.fromDate(DateTime(2020, 11, 16));
  print(
      'The week number of my birthday in 2020 was ${birthdayWeek.weekNumber}');
  //OUTPUT: The week number of my birthday in 2020 was 47

  Week weekFromIso = Week.fromISOString('2021W25');
  print('Week from ISO string: $weekFromIso');
  //OUTPUT: Week from ISO: 2021W25

  DateTime firstDay = weekFromIso.day(0);
  print('The Week $weekFromIso starts with $firstDay');
  //OUTPUT: The Week 2021W25 starts with 2021-06-21 00:00:00.000
}
```


CHANGELOG.md
## 1.1.4
* Add dependabot
* Update dependencies and use fleasy for analysis_options
* Fix analysis suggestions

## 1.1.3

* Use flutter_lints instead of pedantic (deprecated)
* Add badges to README

## 1.1.2

* Make DateTime extension private to avoid conflicts

## 1.1.1

* Fix daylight-saving issue

## 1.1.0

* Stable null safety release
* Add linting rules
* Add tests
* Update examples

## 1.1.0-nullsafety.0

* Add null safety
* Upgrade Dart SDK constraints to >=2.12.0-0 <3.0.0

## 1.0.1

* Add example.dart
* Add repository and issue tracker

## 1.0.0 

* Initial release




test/isoweek_test.dart
import 'package:isoweek/isoweek.dart';
import 'package:test/test.dart';

void main() {
  group('Test conversion from ISO String to week.', () {
    test(
      '2021W07',
      () {
        final weekFromIso = Week.fromISOString('2021W07');

        expect(weekFromIso, equals(const Week(year: 2021, weekNumber: 7)));
      },
    );
    test(
      '2021W7',
      () {
        final weekFromIso = Week.fromISOString('2021W7');

        expect(weekFromIso, equals(const Week(year: 2021, weekNumber: 7)));
      },
    );
    test(
      '2021-W01',
      () {
        final weekFromIso = Week.fromISOString('2021-W01');

        expect(weekFromIso, equals(const Week(year: 2021, weekNumber: 1)));
      },
    );
    test(
      '2021-W1',
      () {
        final weekFromIso = Week.fromISOString('2021-W1');

        expect(weekFromIso, equals(const Week(year: 2021, weekNumber: 1)));
      },
    );
  });

  test(
    'Test getting week from DateTime.',
    () {
      final week = Week.fromDate(DateTime.utc(2020, 11, 16));

      expect(week, equals(const Week(year: 2020, weekNumber: 47)));
    },
  );

  test(
    'Test getting DateTime from week and then getting the DateTime again from the week.',
    () {
      final monday = Week.current().day(0);
      final monday2 = Week.fromDate(monday).day(0);

      expect(monday, equals(monday2));
    },
  );

  test(
    'Test if all DateTimes in the days getter have a time of 00:00:00.',
    () {
      final daysWithTimeEqualToZero = Week.current().days.where(
          (date) => date.hour == 0 && date.minute == 0 && date.second == 0);

      expect(daysWithTimeEqualToZero.length, equals(7));
    },
  );
}


lib/isoweek.dart
library isoweek;

import 'package:equatable/equatable.dart';

class Week extends Equatable {
  const Week({
    required this.year,
    required this.weekNumber,
  });

  /// Return the week that contains the given date.
  factory Week.fromDate(DateTime date) {
    // Add 3 to always compare with January 4th, which is always in week 1
    // Add 7 to index weeks starting with 1 instead of 0
    final woy = (_ordinalDate(date) - date.weekday + 10) ~/ 7;

    // If the week number equals zero, it means that the given date belongs to the preceding (week-based) year.
    if (woy == 0) {
      // The 28th of December is always in the last week of the year
      return Week.fromDate(DateTime(date.year - 1, 12, 28));
    }

    // If the week number equals 53, one must check that the date is not actually in week 1 of the following year
    if (woy == 53 &&
        DateTime(date.year, 1, 1).weekday != DateTime.thursday &&
        DateTime(date.year, 12, 31).weekday != DateTime.thursday) {
      return Week(year: date.year + 1, weekNumber: 1);
    }

    return Week(year: date.year, weekNumber: woy);
  }

  /// Return the current week.
  factory Week.current() => Week.fromDate(DateTime.now());

  /// Return a week initialized from an ISO formatted string like "2020W01" or "2020-W01" (or even "2020-W1" or "2020W1").
  factory Week.fromISOString(String isoString) {
    final parts = isoString.split('W');

    var yearStr = parts.first;
    if (yearStr.length > 4) {
      // If separated with - (e.g. "2020-W01")
      yearStr = yearStr.substring(0, 4);
    }
    final weekNumberStr = parts[1];

    final year = int.parse(yearStr);
    final weekNumber = int.parse(weekNumberStr);

    return Week(year: year, weekNumber: weekNumber);
  }

  final int year;
  final int weekNumber;

  @override
  List<Object> get props => [year, weekNumber];

  /// Return the ordinal date, the number of days since December 31st the previous year.
  static int _ordinalDate(DateTime date) {
    const offsets = [0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334];

    return offsets[date.month - 1] +
        date.day +
        (_isLeapYear(date.year) && date.month > 2 ? 1 : 0);
  }

  /// Return true if the given year is a leap year.
  static bool _isLeapYear(int year) {
    return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
  }

  /// Return the week that many number of weeks into the future.
  Week addWeeks(int weeks) {
    return Week.fromDate(day(0).addWeeks(weeks));
  }

  /// Return the week that many number of weeks in the past.
  Week subtractWeeks(int weeks) {
    return Week.fromDate(day(0).subtractWeeks(weeks));
  }

  /// Return a day of the week as a DateTime object based on the provided day index. Day 0 is Monday.
  DateTime day(int day) {
    // According to ISO the Jan 4th must be in week 1
    final addDays =
        (weekNumber - 1) * 7 + (-DateTime(year, 1, 4).weekday + day) + 1;

    return DateTime(year, 1, 4).addDays(addDays);
  }

  /// Return the next week.
  Week get next => addWeeks(1);

  /// Return the week before.
  Week get previous => subtractWeeks(1);

  /// Return a list of all days in the week.
  List<DateTime> get days => List.generate(7, (i) => day(0).addDays(i));

  /// Return an ISO formatted string like "2020W01".
  @override
  String toString() => '${year}W${weekNumber.toString().padLeft(2, '0')}';
}

extension _DateTimeUtils on DateTime {
  /// Alternative to add(days: x) because of daylight-saving issues.
  DateTime addDays(int days) => copyWith(day: day + days);

  /// Alternative to add(days: x * 7) because of daylight-saving issues.
  DateTime addWeeks(int weeks) => copyWith(day: day + weeks * 7);

  /// Alternative to subtract(days: x * 7) because of daylight-saving issues.
  DateTime subtractWeeks(int weeks) => copyWith(day: day - weeks * 7);

  DateTime copyWith({
    int? year,
    int? month,
    int? day,
    int? hour,
    int? minute,
    int? second,
    int? millisecond,
    int? microsecond,
  }) {
    return DateTime(
      year ?? this.year,
      month ?? this.month,
      day ?? this.day,
      hour ?? this.hour,
      minute ?? this.minute,
      second ?? this.second,
      millisecond ?? this.millisecond,
      microsecond ?? this.microsecond,
    );
  }
}


example/example.dart
import 'package:isoweek/isoweek.dart';

void main() {
  Week currentWeek = Week.current();
  print('Current week: $currentWeek');
  //OUTPUT: 2021W13

  print('Days of current week: ${currentWeek.days}');
  //OUTPUT: [2021-03-29 00:00:00.000, 2021-03-30 00:00:00.000, ...]

  Week nextWeek = currentWeek.next;
  print('Days of next week: ${nextWeek.days}');
  //OUTPUT: Days of next week: [2021-04-05 00:00:00.000, 2021-04-06 00:00:00.000, ...]

  Week previousWeek = currentWeek.previous;
  print('Days of previous week: ${previousWeek.days}');
  //OUTPUT: Days of previous week: [2021-03-22 00:00:00.000, 2021-03-23 00:00:00.000, ...]

  Week futureWeek = currentWeek.addWeeks(5);
  print('5 weeks ahead: $futureWeek');
  //OUTPUT: 5 weeks ahead: 2021W18

  Week birthdayWeek = Week.fromDate(DateTime(2020, 11, 16));
  print(
      'The week number of my birthday in 2020 was ${birthdayWeek.weekNumber}');
  //OUTPUT: The week number of my birthday in 2020 was 47

  Week weekFromIso = Week.fromISOString('2021W25');
  print('Week from ISO string: $weekFromIso');
  //OUTPUT: Week from ISO: 2021W25

  DateTime firstDay = weekFromIso.day(0);
  print('The Week $weekFromIso starts with $firstDay');
  //OUTPUT: The Week 2021W25 starts with 2021-06-21 00:00:00.000
}
